Classes/common/CDTIncrementalStore.m

```diff
@@ -0,0 +1,2409 @@
+}
+
+- (BOOL)replicateInDirection:(CDTISReplicateDirection)direction
+                   withError:(NSError **)error
+                withProgress:(CDTISProgressBlock)progress;
+{
+    if (direction == push) {
+        return [self pushToRemote:error withProgress:progress];
+    }
+    return [self pullFromRemote:error withProgress:progress];
+}
+
+/**
+ *  configure the replicators
+ *
+ *  @param remoteURL remoteURL
+ *  @param manager   manager
+ *  @param datastore datastore
+ *
+ *  @return YES/NO. If `NO` then the caller should continue with local database
+ *          only.
+ */
+- (BOOL)setupReplicators:(NSURL *)remoteURL
+                 manager:(CDTDatastoreManager *)manager
+               datastore:(CDTDatastore *)datastore
+{
+    NSError *err = nil;
+
+    // If remoteURL has a host component, then we have a replication target
+    if (![remoteURL host]) {
+        CDTLogInfo(CDTREPLICATION_LOG_CONTEXT, @"%@: no host component, so no replication",
+                   CDTISType);
+        return NO;
+    }
+
+    NSString *clean = [self cleanURL:remoteURL];
+
+    CDTReplicatorFactory *repFactory =
+        [[CDTReplicatorFactory alloc] initWithDatastoreManager:manager];
+    if (!repFactory) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create replication factory",
+                    CDTISType, clean);
+        return NO;
+    }
+
+    CDTPushReplication *pushRep =
+        [CDTPushReplication replicationWithSource:datastore target:remoteURL];
+    if (!pushRep) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create push replication object",
+                    CDTISType, clean);
+        return NO;
+    }
+
+    CDTReplicator *pusher = [repFactory oneWay:pushRep error:&err];
+    if (!pusher) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create replicator for push: %@",
+                    CDTISType, clean, err);
+        return NO;
+    }
+
+    CDTPullReplication *pullRep =
+        [CDTPullReplication replicationWithSource:remoteURL target:datastore];
+    if (!pullRep) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create pull replication object",
+                    CDTISType, clean);
+        return NO;
+    }
+
+    CDTReplicator *puller = [repFactory oneWay:pullRep error:&err];
+    if (!puller) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create replicator for pull: %@",
+                    CDTISType, clean, err);
+        return NO;
+    }
+
+    self.remoteURL = remoteURL;
+
+    self.puller = puller;
+    puller.delegate = self;
+
+    self.pusher = pusher;
+    pusher.delegate = self;
+
+    return YES;
+}
+
+- (BOOL)linkReplicators:(NSURL *)remoteURL
+{
+    return [self setupReplicators:remoteURL manager:self.manager datastore:self.datastore];
+}
+
+- (void)unlinkReplicators
+{
+    self.pusher = nil;
+    self.puller = nil;
+}
+
+static NSString *fixupName(NSString *name)
```

Comment:
We should always check compliance and report the failure adivising the switch.
Should doc this as well.